etools-searchable-multiselection-menu.html

```diff
@@ -391,56 +438,130 @@
           }
         },
         _valueChanged: function(value) {
-          if (this.multi && value && value.length > 0 ) {
+          if (!value) {
+            // empty value/reset selection
+            this.set('selectedValues', null);
+            return;
+          }
+          if (this.multi && Array.isArray(value) && value.length > 0 ) {
+            // multiple selection on
             this.set('selectedValues', value.map(function(item) {
-              return '' + item.value;
-            }));
+              return '' + item[this.optionValue];
+            }.bind(this)));
           } else if (!this.multi && Array.isArray(value) && value.length > 0) {
             /**
              * In case the single selection dropdown is initialized with an array of one object
              * value and selectedValues must be updated with that object data
              */
             this.set('value', value[0]);
-            this.set('selectedValues', [value[0].value]);
-          } else if (!value) {
-            this.set('selectedValues', null);
+            this.set('selectedValues', [value[0][this.optionValue]]);
           }
         },
-        _valueSelected: function() {
-          var self = this;
-          if(this.selectedValues === null) {
-            this.set('value', null);
-            this._fireValueChangeEvent();
+        // get esmm single selection selected label to be displayed
+        _getSingleSelectedOptionLabel: function(value) {
+          if (typeof value === 'object' && value[this.optionLabel]) {
+            return value[this.optionLabel];
           }
-          if (this.selectedValues instanceof Array) {
-            var newValue = [];
-            this.selectedValues.forEach(function(selected) {
-              self.options.forEach(function(option) {
-                if (parseInt(option.value, 10) === parseInt(selected, 10)) {
-                  newValue.push(option);
+          return null;
+        },
+        _valueSelected: function() {
```

Comment:
I think this is now too big and we need to split it in two functions, and leave here only the check to determine if we are using single selection or multi selection